lib/streamlit/connections/sql_connection.py

```diff
@@ -0,0 +1,226 @@
+
+        Aside from the `ttl` kwarg, all kwargs passed to this function are passed down
+        to [pd.read_sql](https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html).
+        and have the behavior described in the pandas documentation.
+
+        Returns
+        -------
+        pd.DataFrame
+            The result of running the query, formatted as a pandas DataFrame.
+        """
+
+        from sqlalchemy import text
+        from sqlalchemy.exc import DatabaseError, InternalError, OperationalError
+        from tenacity import (
+            retry,
+            retry_if_exception_type,
+            stop_after_attempt,
+            wait_fixed,
+        )
+
+        @retry(
+            after=lambda _: self.reset(),
+            stop=stop_after_attempt(3),
+            reraise=True,
+            retry=retry_if_exception_type(
+                (DatabaseError, InternalError, OperationalError)
+            ),
+            wait=wait_fixed(1),
+        )
+        @cache_data(
+            show_spinner="Running `sql.query(...)`.",
+            ttl=ttl,
+        )
+        def _query(
+            sql: str,
+            index_col=None,
+            chunksize=None,
+            params=None,
+            **kwargs,
+        ) -> pd.DataFrame:
+            instance = self._instance.connect()
+            return pd.read_sql(
+                text(sql),
+                instance,
+                index_col=index_col,
+                chunksize=chunksize,
+                params=params,
+                **kwargs,
+            )
+
+        return _query(
+            sql,
+            index_col=index_col,
+            chunksize=chunksize,
+            params=params,
+            **kwargs,
+        )
+
+    @contextmanager
+    def session(self) -> Iterator["Session"]:
+        """A thin wrapper around SQLAlchemy Session context management.
+
+        This allows us to write
+            `with conn.session() as session:`
+        instead of importing the sqlalchemy.orm.Session object and writing
+            `with Session(conn._instance) as session:`
+
+        Users of this connection should use the contextmanager pattern for writes,
+        transactions, and anything more complex than simple read queries.
+
+        See the usage example below, which assumes we have a table `numbers` with a
+        single integer column `val`. The [SQLAlchemy](https://docs.sqlalchemy.org/en/20/orm/session_basics.html)
+        docs also contain much more information on the usage of sessions.
+
+        Example
+        -------
+        >>> n = st.slider("Pick a number")
+        >>> if st.button("Add the number!"):
+        ...     with conn.session() as session:
+        ...         session.execute("INSERT INTO numbers (val) VALUES (:n);", {"n": n})
+        ...         session.commit()
+        """
+        from sqlalchemy.orm import Session
+
+        with Session(self._instance) as s:
+            yield s
```

Comment:
The Session object has its own context mgr support, right?

If so, do we gain anything by wrapping this with an additional with / yield? It seems to make it less flexible (get an error if I call `.Session()` outside `with:`).

Seems like the current code still works without it, and it also supports calling outside a with block.

Reply:
Ah, if this works fine just returning the `Session()` rather than yielding it, then I think we should do that. I wasn't sure when writing it whether it was valid and must have forgotten to check

Reply:
Yeah this seems to work fine:

```python
conn = st.experimental_connection('pets_db', type='sql')

from sqlalchemy.orm import Session
with Session(conn._instance) as s:
    s.execute('CREATE TABLE IF NOT EXISTS pet_owners (person TEXT, pet TEXT);')
    s.execute('DELETE FROM pet_owners;')
    pet_owners = {'jerry': 'fish', 'barbara': 'cat', 'alex': 'puppy'}
    for k in pet_owners:
        s.execute(
            'INSERT INTO pet_owners (person, pet) VALUES (:owner, :pet);',
            params=dict(owner=k, pet=pet_owners[k])
        )
    s.commit()

df = conn.query("select * from pet_owners;", ttl=0)
st.dataframe(df)
```

And same code works if I just initialize the session like that ^ outside the `with:` block.